src/contracts/core/KeyRegistrar.sol

```diff
@@ -0,0 +1,439 @@
+        } 
+        else if (config.curveType == CurveType.BN254) {
+            _registerBN254Key(operatorSet, operator, pubkey, signature);
+        } 
+        else {
+            revert InvalidCurveType();
+        }
+
+        emit KeyRegistered(operatorSet, operator, config.curveType, pubkey);
+    }
+
+    /// @inheritdoc IKeyRegistrar
+    function deregisterKeyAndUpdate(
+        address operator,
+        OperatorSet memory operatorSet
+    ) external checkCanCall(operatorSet.avs) {
+        
+        OperatorSetConfig memory config = operatorSetConfigs[operatorSet.key()];
+        require(config.isActive, OperatorSetNotConfigured());
+        
+        KeyInfo memory keyInfo = operatorKeyInfo[operatorSet.key()][operator];
+        
+        if (!keyInfo.isRegistered) {
+            // Revert if key was not registered
+            revert KeyNotFound(operatorSet, operator);
+        }
+
+        if (config.curveType == CurveType.BN254) {
+
+            // Extract the G1 point from the stored key data
+            (uint256 g1X, uint256 g1Y, , ) = abi.decode(keyInfo.keyData, (uint256, uint256, uint256[2], uint256[2]));
+            BN254.G1Point memory operatorKey = BN254.G1Point(g1X, g1Y);
+
+            // Remove the operator's key from the aggregate
+            _updateAggregateBN254Key(operatorSet, operatorKey, false);
+        }
+
+        // Clear key info
+        delete operatorKeyInfo[operatorSet.key()][operator];
+
+        emit KeyDeregistered(operatorSet, operator, config.curveType);
+    }
+
+    /// @inheritdoc IKeyRegistrar
+    function checkAndUpdateKey(
+        OperatorSet memory operatorSet,
+        address operator
+    ) external checkCanCall(operatorSet.avs) {
+        OperatorSetConfig memory config = operatorSetConfigs[operatorSet.key()];
+        require(config.isActive, OperatorSetNotConfigured());
+        
+        KeyInfo memory keyInfo = operatorKeyInfo[operatorSet.key()][operator];
+        if (!keyInfo.isRegistered) {
+            revert KeyNotFound(operatorSet, operator);
+        }
+
+        if (config.curveType == CurveType.BN254) {
+            // Extract the G1 point from the stored key data
+            (uint256 g1X, uint256 g1Y, , ) = abi.decode(keyInfo.keyData, (uint256, uint256, uint256[2], uint256[2]));
+            BN254.G1Point memory operatorKey = BN254.G1Point(g1X, g1Y);
+
+            // Add the operator's key to the aggregate
+            _updateAggregateBN254Key(operatorSet, operatorKey, true);
+        }
+    }
+
+    /**
+     *
+     *                         INTERNAL FUNCTIONS
+     *
+     */
+
+    /**
+     * @notice Validates and registers an ECDSA public key
+     * @param operatorSet The operator set to register the key for
+     * @param operator Address of the operator
+     * @param pubkey The ECDSA public key bytes
+     * @dev Validates key format and ensures global uniqueness
+     */
+    function _registerECDSAKey(
```

Comment:
should verify a sig over a message that includes the operator's address, no? otherwise, anyone can frontrun to burn keys and cause ECDSA registrations to fail